### PR TITLE
Introduce LauncherConfig for configuring engines and listeners

### LINKED ISSUE #767: Allow LauncherFactory to be configured regarding engines and listeners
## Overview

Currently the only way to create DefaultLauncher is use `LauncherFactory.create()` method, that loads engines and listeners via SPI. In some cases it would be nice to have an ability to directly specify the engines and listeners for launcher. 

In my case I want to load engines with SPI, but disable SPI for listeners

### PR DESCRIPTION
**Pull request by baev**
## Overview

The pull request addresses the feature request in #767.

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests)
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [ ] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by sbrannen** [comment]
Currently slated for 5.1 RC1 in order to discuss amongst the team prior to 5.1 GA.

**Comment by sbrannen** [comment]
Oops: you broke the build.

Please apply Spotless: https://travis-ci.org/junit-team/junit5/jobs/334739803#L1411

**Comment by baev** [comment]
okey.

BTW should I add `throws PreconditionViolationException` to all the methods or remove it from all the methods?

**Comment by sbrannen** [comment]
> BTW should I add `throws PreconditionViolationException` to all the methods or remove it from all the methods? 

Please leave them in place.

**Comment by marcphilipp** [comment]
@baev Can you please explain your use case for each of the new methods?

**Comment by baev** [comment]
In general we need to be able to create Launcher without loading listeners via SPI to test our custom listener. In our tests we run sample tests (test with tags, dynamic tests and so on) using launcher and then checking stored data. 

At the moment the problem is that main listener (not a tested one) is being loaded as well as the tested one. We only need a `createWithoutLoadingListeners` method, other are created to make API more completed.

I can see this methods can be helpful for testing custom engines as well. And also for custom test executors systems (as far as I know there are lot of usage of `JUnitCore` API in the enterprise world, so I guess `LaunchFactory` API will have much use as well).

**Comment by marcphilipp** [comment]
Thanks for the explanation! I think for listeners this does make sense.

For custom engines, you can already pass an `EngineFilter` as part of the `LauncherDiscoveryRequest` to only include the engine you're interested in.

Implementation-wise we have to decide whether we want this to be a builder because as soon as we add another feature we'd have to double the number of factory methods.

**Comment by baev** [comment]
> Implementation-wise we have to decide whether we want this to be a builder because as soon as we add another feature we'd have to double the number of factory methods.

Yeah, agree. At least we should make a SPI-loaders public so people can use it in case of custom `Launcher`. And then having only few factory methods/constructors would be enough (all args one and required args one).

Also I'm not sure what is the point of having `DefaultLauncher` package private.

**Comment by marcphilipp** [comment]
> Also I'm not sure what is the point of having `DefaultLauncher` package private.

We want people to rely on the `Launcher` interface instead of on the implementation. Thus, we hide it this way.

**Comment by baev** [comment]
Yeah, but anyway to use a launcher people should use factory that will return `Launcher` interface. 

I want to make use of DefaultLauncher at least to check that my listener is loaded via SPI (like you did in your tests, check that listener registry contains the listener after Launcher is created).

BTW this topic not hits the scope of the pull request, just thoughts

**Comment by jbduncan** [comment]
@Zir0-93 I'm a bit confused by the message produced by @clarity-bot. The message it produced suggests that it opened this PR, whereas I'm pretty sure that @baev opened it. Can you clarify things for me? 🤔

**Comment by marcphilipp** [comment]
@junit-team/junit-lambda Did one of you enable claritybot?

**Comment by marcphilipp** [comment]
Hmm, apparently it was @Zir0-93...

**Comment by Zir0-93** [comment]
Sorry for any confusion this caused, I will definitely work on cleaning his message up! To answer your question, no it is not enabled. Clarity-bot is just posting a sample structure-diff on this Pull Request to try and teams interested, if you do nothing you will never see him again!

**Comment by marcphilipp** [comment]
Instead of adding a bunch of overloaded methods, I propose the following:

- Add `LauncherFactory.create(LauncherConfig)`.
- Add `LauncherConfig` interface
- Add `LauncherConfig.builder()` to build an instance of `LauncherConfig` with auto registration enabled by default and methods like `enableTestExecutionListenerAutoRegistration()` etc.

```java
interface LauncherConfig {
    boolean isTestExecutionListenerAutoRegistrationEnabled()
    boolean isTestEngineAutoRegistrationEnabled()
    Collection<TestEngine> getAdditionalTestEngines()
    Collection<TestExecutionListener> getAdditionalTestExecutionListeners()
}
```

@baev @junit-team/junit-lambda Thoughts?

**Comment by baev** [comment]
sounds good to me

**Comment by baev** [comment]
@marcphilipp should I update the pull request?

**Comment by marcphilipp** [comment]
@baev I haven't had a chance to discuss it in the team, yet. Thus, there's a chance I change my mind. On the other hand, it's always easier to discuss concrete code than an somewhat abstract proposal. So, if you want, feel free to go ahead and update the PR.

**Comment by marcphilipp** [comment]
**Team Decision:** Let's implement the above proposal.

@baev Feel free to work on this one. 🙂

**Comment by baev** [comment]
Will try to find some time next week

**Comment by baev** [comment]
@marcphilipp added requested changes. If code looks ok, I'll add tests and javadocs

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+			return this;
+		}
+
+		public Builder addAdditionalTestEngines(TestEngine... engines) {
+			this.additionalTestEngines.addAll(Arrays.asList(engines));
+			return this;
+		}
+
+		public Builder addAdditionalTestExecutionListeners(TestExecutionListener... listeners) {
+			this.additionalTestExecutionListeners.addAll(Arrays.asList(listeners));
```
Comment (marcphilipp) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java]: Please use `Collections.addAll(Collection<? super T>, T...)` instead.

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+
+import org.apiguardian.api.API;
+import org.junit.platform.engine.TestEngine;
+import org.junit.platform.launcher.TestExecutionListener;
+
+@API(status = EXPERIMENTAL, since = "1.3.0")
+interface LauncherConfig {
```
Comment (marcphilipp) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java]: The interface should be `public`.

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+
+import org.apiguardian.api.API;
+import org.junit.platform.engine.TestEngine;
+import org.junit.platform.launcher.TestExecutionListener;
+
+@API(status = EXPERIMENTAL, since = "1.3.0")
```
Comment (marcphilipp) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java]: We use `1.3` unless the micro version is not zero.

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+	class Builder {
+
+		private boolean testExecutionListenerAutoRegistrationEnabled = true;
+
+		private boolean testEngineAutoRegistrationEnabled = true;
+
+		private Collection<TestEngine> additionalTestEngines = new ArrayList<>();
+
+		private Collection<TestExecutionListener> additionalTestExecutionListeners = new ArrayList<>();
```
Comment (marcphilipp) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java]: 2x `final`

**Comment by baev** [comment]
@marcphilipp done

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncherConfig.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import static java.util.Collections.unmodifiableCollection;
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+import java.util.Collection;
+
+import org.apiguardian.api.API;
+import org.junit.platform.engine.TestEngine;
+import org.junit.platform.launcher.TestExecutionListener;
+
+@API(status = EXPERIMENTAL, since = "1.3.0")
```
Comment (sbrannen) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncherConfig.java]: Non-public types do not need an `@API` declaration.

Please remove.

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncherConfig.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+	private final boolean testEngineAutoRegistrationEnabled;
+
+	private final Collection<TestEngine> additionalTestEngines;
+
+	private final boolean testExecutionListenerAutoRegistrationEnabled;
+
+	private final Collection<TestExecutionListener> additionalTestExecutionListeners;
+
+	DefaultLauncherConfig(final boolean testEngineAutoRegistrationEnabled,
```
Comment (sbrannen) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncherConfig.java]: We don't declare constructor parameters as `final` unless necessary.

Please remove the `final` declarations.

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+		private final Collection<TestExecutionListener> additionalTestExecutionListeners = new ArrayList<>();
+
+		/**
+		 * Configure test execution listeners auto detection.
+		 *
+		 * @param enabled true if test execution listeners should be auto detected and false otherwise.
+		 * @return this builder for method chaining
+		 */
+		public Builder setTestExecutionListenerAutoRegistration(boolean enabled) {
```
Comment (sbrannen) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java]: Rename method to `setTestExecutionListenerAutoRegistrationEnabled()`.

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+ *
+ * <h4>Example</h4>
+ *
+ * <pre class="code">
+ * import static org.junit.platform.launcher.core.*;
+ * // ...
+ *
+ *   LauncherConfig.builder()
+ *     .setTestEngineAutoRegistrationEnabled(true)
+ *     .setTestExecutionListenerAutoRegistration(true)
```
Comment (sbrannen) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java]: Rename method to `setTestExecutionListenerAutoRegistrationEnabled()`.

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherFactory.java [hunk 9]
```diff
@@ -34,8 +39,8 @@
  * {@link Launcher#registerTestExecutionListeners(TestExecutionListener...)}
  * method on the created launcher instance.
  *
- * @since 1.0
  * @see Launcher
+ * @since 1.0
```
Comment (sbrannen) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherFactory.java]: Why did you move the `@since` tag?

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherFactory.java [hunk 10]
```diff
@@ -54,10 +59,29 @@ private LauncherFactory() {
...
 	public static Launcher create() throws PreconditionViolationException {
-		Launcher launcher = new DefaultLauncher(new ServiceLoaderTestEngineRegistry().loadTestEngines());
-		for (TestExecutionListener listener : new ServiceLoaderTestExecutionListenerRegistry().loadListeners()) {
-			launcher.registerTestExecutionListeners(listener);
+		return create(LauncherConfig.builder().build());
+	}
+
+	/**
+	 * Factory method for creating a new {@link Launcher} using dynamically
+	 * detected test engines.
```
Comment (sbrannen) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherFactory.java]: That's no longer true. Test engines will only be auto-registered if the supplied config specifies that.

Please reword.

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherFactory.java [hunk 11]
```diff
@@ -54,10 +59,29 @@ private LauncherFactory() {
...
+	 *
```
Comment (sbrannen) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherFactory.java]: Missing a `@param` declaration for the `LauncherConfig`.

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherFactory.java [hunk 12]
```diff
@@ -54,10 +59,29 @@ private LauncherFactory() {
...
+	 * @throws PreconditionViolationException if no test engines are detected
+	 */
+	@API(status = EXPERIMENTAL, since = "1.3")
+	public static Launcher create(LauncherConfig config) throws PreconditionViolationException {
+		List<TestEngine> engines = new ArrayList<>();
```
Comment (sbrannen) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherFactory.java]: Missing a precondition "non-null" check for the supplied config.

**Comment by marcphilipp** [comment]
@baev Do you have time to address the changes proposed by @sbrannen or shall we take it from here?

**Comment by baev** [comment]
I’m good to handle requested changes next week, on work trip atm

**Comment by baev** [comment]
@sbrannen done

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+ *     .setTestEngineAutoRegistrationEnabled(true)
+ *     .setTestExecutionListenerAutoRegistrationEnabled(true)
+ *     .addAdditionalTestEngines(new CustomTestEngine())
+ *     .addAdditionalTestExecutionListeners(new CustomTestExecutionListener())
+ *     .build();
+ * </pre>
+ *
+ * @see Launcher
+ * @see LauncherFactory
+ * @since 1.0
```
Comment (sbrannen) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java]: This should be `@since 1.3`.

Comment (sbrannen) [reply junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java]: But I think we'll just fix that after merging into `master`.

**Comment by sbrannen** [comment]
This has been merged into `master`.

Thanks for the contribution!

**Comment by sbrannen** [comment]
Reopening to address the remaining TODOs:

- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [ ] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
